lib/utils.coffee

```diff
@@ -128,30 +128,39 @@ distanceForRange = ({start, end}) ->
   column = end.column - start.column
   new Point(row, column)
 
-getNewTextRangeFromChanges = (changes) ->
+getNewTextRangeFromUndoStack = (stack) ->
   finalRange = null
-  for change in changes when change.newRange?
-    {oldRange, oldText, newRange, newText} = change
-    unless finalRange?
-      finalRange = newRange.copy() if newText.length
-      continue
-    # shrink
-    if oldText.length and finalRange.containsRange(oldRange)
-      amount = oldRange
-      diff = distanceForRange(amount)
-      diff.column = 0 unless (amount.end.row is finalRange.end.row)
-      finalRange.end = finalRange.end.translate(diff.negate())
-    # extend
-    if newText.length and finalRange.containsPoint(newRange.start)
-      amount = newRange
-      diff = distanceForRange(amount)
-      diff.column = 0 unless (amount.start.row is finalRange.end.row)
-      finalRange.end = finalRange.end.translate(diff)
+  for patch in stack when patch.getChanges?
+    for change in patch.getChanges()
+      {oldStart, oldExtent, oldText, newStart, newExtent, newText} = change
+      oldRange = new Range(
+        [oldStart.row, oldStart.column],
+        [oldStart.row + oldExtent.row, oldStart.column + oldExtent.column]
+      )
+      newRange = new Range(
+        [newStart.row, newStart.column],
+        [newStart.row + newExtent.row, newStart.column + newExtent.column]
+      )
+      unless finalRange?
+        finalRange = newRange if newText.length
+        continue
+      # shrink
+      if oldText.length and finalRange.containsRange(oldRange)
+        amount = oldRange
+        diff = distanceForRange(amount)
+        diff.column = 0 unless (amount.end.row is finalRange.end.row)
+        finalRange.end = finalRange.end.translate(diff.negate())
+      # extend
+      if newText.length and finalRange.containsPoint(newRange.start)
+        amount = newRange
+        diff = distanceForRange(amount)
+        diff.column = 0 unless (amount.start.row is finalRange.end.row)
+        finalRange.end = finalRange.end.translate(diff)
   finalRange
```

Comment:
Instead of manually merging finalRange like above, can we use `Patch.compose(changes)` to let Patch calculate finalRange?
If so, we can eliminate all of above code, thats would be great!